src/book/02-projects/01-nubots/03-build-system.mdx

```diff
@@ -5,3 +5,526 @@ title: Build System
+
+The hierarchical structure is as follows
+
+```text
+config/
+├── nugus1/
+│   └── ExampleModule.yaml
+├── nugus2/
+│   └── ExampleModule.yaml
+├── robocup/
+│   └── ExampleModule.yaml
+└── ExampleModule.yaml
+```
+
+The file `config/ExampleModule.yaml` must exist for every module. This file defines all of the configuration for the module.
+`config/nugus1/ExampleModule.yaml` is an optional file and allows you to provide robot-specific overrides for specific configuration values, in this case it would provide overrides for the `nugus1` robot.
+Similarly, `config/robocup/ExampleModule.yaml` allows for the `robocup` role to further override configuration values.
+
+It is intended that all configuration values exist in the default configuration file and that only a subset of configuration values appear in the robot-specific and role-specific override files.
+
+The robot-specific override is intended to allow configuration values to be tweaked to account a slight variation in the robots hardware. For instance, the cameras on each robot have different serial numbers.
+
+The role-specific override is intended to change a modules behaviour based on which role is running. For instance, disabling the transmission of certain NUsight2 traffic while playing a competition game.
+
+To reiterate how the configuration system loads and merges configuration files
+
+1. The default configuration file `config/ModuleName.yaml` is loaded. This forms the base configuration for the module
+1. A check is made to see if `config/RobotName/ModuleName.yaml` exists. If it does, it is loaded and all values found therein override the equivalent values in the base configuration
+1. A check is made to see if `config/RoleName/ModuleName.yaml` exists. If it does, it is loaded and all values found therein override the equivalent values in the base configuration
+1. If any of these 3 configuration files are modified then 1-3 are run again
+
+<Alert type='info'>
+  It is permitted for a module to not have a configuration file. However, if the
+  module does have configuration, the above rules need to be adhered to.
+</Alert>
+
+## The Script System
+
+The script system follows a similar structure as the configuration system. However, default scripts are moved to a platform type folder and the overriding system is not used as it doesn't make much sense to try and merge two scripts.
+
+In the script system, we first check for a robot-specific script. If a robot-specific script does not exist, then we check for a platform-specific script.
+
+```text
+scripts/
+├── nugus/
+│   └── Stand.yaml
+└── nugus2/
+    └── Stand.yaml
+```
+
+To reiterate how the script system loads script files
+
+1. A check is made to see if `config/RobotName/Script.yaml` exists. If it does, it is loaded
+1. If it doesn't exist, then a check is made to see if `config/PlatformName/Script.yaml` exists. If it does, it is loaded
+
```

Comment:
These two really don't seem to be a part of the build system... Or at least, the aspects of them discussed here seem to stray beyond the build system

Reply:
Pretty important information though. Perhaps we could have a **Configuration System** page?

Reply:
Oh 100% we should keep this information somewhere. It’s just that if I were looking for this information, “Build System” is not where I would look.

Reply:
A **Configuration and Scripts** page that includes those systems as well as script tuner/runner/engine? Or two pages, **Configuration** and **Scripts**? Though **Configuration** would be pretty short?

Reply:
Added PR #40 for this